src/os/bluestore/BlueStore.cc

```diff
@@ -13901,13 +13926,20 @@ void BlueStore::_do_write_small(
   BlobRef b = c->new_blob();
   uint64_t b_off = p2phase<uint64_t>(offset, alloc_len);
   uint64_t b_off0 = b_off;
-  _pad_zeros(&bl, &b_off0, block_size);
-  o->extent_map.punch_hole(c, offset, length, &wctx->old_extents);
-  wctx->write(offset, b, alloc_len, b_off0, bl, b_off, length,
-    min_alloc_size != block_size, // use 'unused' bitmap when alloc granularity
-                                  // doesn't match disk one only
-    true);
-
+  // Zero detection -- if the bufferlist contains only zeroes, we logically extend
+  // the space without writing the zeroed bufferlist to the disk for better optimization.
+  if (!bl.is_zero()) {
+    // There is relevant metadata in this bufferlist, so we pad zeroes,
+    // extend the space, and write to disk.
+    _pad_zeros(&bl, &b_off0, block_size);
+    o->extent_map.punch_hole(c, offset, length, &wctx->old_extents);
+    wctx->write(offset, b, alloc_len, b_off0, bl, b_off, length,
+        min_alloc_size != block_size, // use 'unused' bitmap when alloc granularity doesn't match disk one only
+        true);
+  } else { // if (bl.is_zero())
+    // Logically extend the space without writing to disk.
+    o->extent_map.punch_hole(c, offset, length, &wctx->old_extents);
+  }
```

Comment:
Given the same pattern to perform zero detection and to apply the write 
I would suggest to introduce a wrapper method to avoid code duplication...

Reply:
Agreed. I want to leave this for now while I work on fixing the fine-grain implementation, but this will be a fix that will for sure happen once the fine-grain is fixed.

I really appreciate the thorough review, @ifed01!

Reply:
> Given the same pattern to perform zero detection and to apply the write
> I would suggest to introduce a wrapper method to avoid code duplication...
@ifed01 
I think that once we got rid of fine-grain zero detection having separate wrappers is more harm then good.
It passes a ton of parameters and provides limited usability.
@ljflores But ultimately, this is your decision to take.

Reply:
> It passes a ton of parameters and provides limited usability.

I see what you mean. The new functions do pass a lot of parameters, which I understand can be confusing.

On the flip-side, it does help to ensure consistency between zero checks and writes, especially in `_do_write_small()` when it happens so often.

Before I go refactoring that part, since it's more of a structural issue, I would like to have the logic and unit tests approved. But perhaps at the end, if we decide that the logic and unit tests look good, but the extra functions are too messy, I can make a last change to refactor the code and write the checks how they were initially. I'm also curious to see what @ifed01 thinks.